rx_signal_processing/rx_dsp_chain.cu

```diff
@@ -238,129 +218,124 @@ int main(int argc, char **argv){
-                             rx_freqs, &filters, beam_phases,
+                             rx_freqs, complex_taps, beam_phases,
                              beam_direction_counts, rx_metadata.initialization_time(),
-                             rx_metadata.sequence_start_time(), slice_ids);
+                             rx_metadata.sequence_start_time(), slice_ids, dm_rates);
 
     if (rx_metadata.numberofreceivesamples() == 0){
       //TODO(keith): handle error for missing number of samples.
     }
 
-    //We need to sample early to account for propagating samples through filters. The number of 
+
+    //We need to sample early to account for propagating samples through filters. The number of
     //required early samples is equal to the largest filter length in time (based on the rate at
-    //that stage.) The following lines find the max filter length in time and convert that to 
+    //that stage.) The following lines find the max filter length in time and convert that to
     //number of samples at the input rate.
-    int64_t extra_samples = (first_stage_dm_rate * second_stage_dm_rate * third_stage_dm_rate *
-                        filters.get_num_fourth_stage_taps());
 
-    if (extra_samples < (first_stage_dm_rate * second_stage_dm_rate * filters.get_num_third_stage_taps())) {
-      extra_samples = first_stage_dm_rate * second_stage_dm_rate * filters.get_num_third_stage_taps();
-    }
+    int64_t extra_samples = std::accumulate(dm_rates.begin(), dm_rates.end()-1, 1,
+                                            std::multiplies<int64_t>()) *
+                            complex_taps.back().size();
 
-    if (extra_samples < (first_stage_dm_rate * filters.get_num_second_stage_taps())) {
-      extra_samples = first_stage_dm_rate * filters.get_num_second_stage_taps();
-    }
+    // problem with this if you have too few stages. Need to figure this out.
 
-    if (extra_samples < (filters.get_num_first_stage_taps())) {
-      extra_samples = filters.get_num_first_stage_taps();
+    for (int32_t i=complex_taps.size()-2; i>=0; i--) {
+      if (i == 0) {
+        if (extra_samples < complex_taps[0].size()) {
+          extra_samples = complex_taps[0].size();
+        }
+        else if (i == 1) {
+          auto dm_index = dm_rates.size() - 2 - i;
+          if (extra_samples < (dm_rates[0] * complex_taps[1].size())) {
+            extra_samples = dm_rates[0] * complex_taps[1].size();
+          }
+        }
+        else {
+          auto dm_index = dm_rates.size() - 2 - i;
+          auto total_dm_rate = std::accumulate(dm_rates.begin(),
+                                            dm_rates.end()-2-i,
+                                            1,
+                                            std::multiplies<uint32_t>());
+          if (extra_samples < total_dm_rate * complex_taps[i].size()) {
+            extra_samples = total_dm_rate * complex_taps[i].size();
+          }
+        }
+      }
     }
```

Comment:
// for (int32_t i=complex_taps.size()-2; i>=0; i--) {

Line 268

Reply:
This whole loop looks clearer and easier to read with the brackets fixed.